toolchain/language_server/language_server.cpp

```diff
@@ -14,20 +14,47 @@
 
 namespace Carbon::LanguageServer {
 
+// An adapter for clangd's logging that sends most messages to `error_stream`.
+// Verbose logging is only printed if `vlog_stream` is provided, and will be
+// sent there.
+class Logger : public clang::clangd::Logger {
+ public:
+  explicit Logger(llvm::raw_ostream* error_stream,
+                  llvm::raw_ostream* vlog_stream)
+      : error_logger_(*error_stream, clang::clangd::Logger::Info),
+        vlog_logger_(vlog_stream
+                         ? std::make_unique<clang::clangd::StreamLogger>(
+                               *vlog_stream, clang::clangd::Logger::Verbose)
+                         : nullptr) {}
+
+  auto log(Level level, const char* format,
+           const llvm::formatv_object_base& message) -> void override {
+    if (level != clang::clangd::Logger::Verbose) {
+      error_logger_.log(level, format, message);
+    } else if (vlog_logger_) {
+      vlog_logger_->log(level, format, message);
+    }
+  }
+
+ private:
+  clang::clangd::StreamLogger error_logger_;
+  std::unique_ptr<clang::clangd::StreamLogger> vlog_logger_;
+};
+
 auto Run(FILE* input_stream, llvm::raw_ostream& output_stream,
-         llvm::raw_ostream& error_stream, FilenameDiagnosticEmitter& emitter)
-    -> bool {
-  // TODO: Consider implementing a custom logger that splits vlog to
-  // vlog_stream when provided. For now, this disables verbose logging.
-  clang::clangd::StreamLogger logger(error_stream, clang::clangd::Logger::Info);
+         llvm::raw_ostream& error_stream, llvm::raw_ostream* vlog_stream,
+         FilenameDiagnosticEmitter& emitter) -> bool {
+  // The language server internally uses diagnostics for logging issues, but the
```

Comment:
Is using diagnostics internally the right thing here?

Specifically, do we need to do anything to avoid internal language server errors showing up in the IDE's UI because we emit them using our diagnostic infrastructure? Maybe the UIs actually handle these diagnostics well? Or maybe there is a way to consume them and print them to the streams instead of sending them back?

Sorry if this is just lack of context on my part -- and not sure it's really blocking, just a bit curious about the strategy here.

Reply:
Right now, this is using the diagnostics system for formatting to print to stderr. I can investigate using the LSP's logMessage if you want, but it might make sense to still use the diagnostics system for formatting.

To be sure, I'm not:

- Trying to make compiler diagnostics show up in the UI. That would be done by the LSP's publishDiagnostics, which we don't support at present.
- Trying to make these internal diagnostics show up in the UI in some way. My best thought is that'd be logMessage. But again, that'd take a bit more work.

Note, in the tests, this is why the diagnostics show up in stderr. If we were transmitting them to the IDE using the LSP, they'd show up as STDOUT and hopefully JSON-formatted.

Reply:
If I'm misunderstanding, please let me know, I'm responding because your suggestion to print to streams felt like it matched the current printing to `error_stream` / `vlog_stream`.

Reply:
Thanks, I think this clarifies.

What led me to the question was not knowing that we're not currently using `publishDiagnostics` (yet). Mostly, I was wondering if this change would risk getting in the way of starting to support that -- I suspect we *don't* want these internal errors to show up using `publishDiagnostics`, and so we'll need to separate them from other kinds of errors. But if we've not gotten to that point yet, happy to wait. If the infra is useful for formatting to the streams, makes sense.

Not pushing for using `logMessage` personally, happy for that to happen or not based on what ends up useful to you or others working here.